classes/event/user_failed_mfa.php

```diff
@@ -0,0 +1,94 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Event for successful MFA authorization.
+ *
+ * @package     tool_mfa
+ * @author      Mikhail Golenkov <golenkovm@gmail.com>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_mfa\event;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Event for when user successfully passed all MFA factor checks.
+ *
+ * @property-read array $other {
+ *      Extra information about event.
+ * }
+ *
+ * @package     tool_mfa
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+class user_failed_mfa extends \core\event\base {
+    /**
+     * Create instance of event.
+     *
+     * @param int $user the User object of the User who failed MFA authentication.
+     *
+     * @return user_failed_mfa the user_passed_mfa event
+     *
+     * @throws \coding_exception
+     */
+    public static function user_failed_mfa_event($user) {
+
+        $data = array(
+            'relateduserid' => null,
+            'context' => \context_user::instance($user->id),
+            'other' => array (
+                'userid' => $user->id
+            )
```

Comment:
can you dump the same stuff that you'd see in the mfa debug mode into this other and then show it in the desc

ie I want to see something like user X failed mfa 'because they didn't know their totp' - can be ugly as sin as long as the data is there